esp32_ulp/opcodes.py

```diff
@@ -634,41 +638,99 @@ def i_jump(target, condition='--'):
     raise TypeError('unsupported operand: %s' % target.raw)
 
 
+def _jump_relr(threshold, cond, offset):
+    """
+    Equivalent of I_JUMP_RELR macro in binutils-esp32ulp
+    """
+    _br.imm = threshold
+    _br.cmp = cond
+    _br.offset = abs(offset)
+    _br.sign = 0 if offset >= 0 else 1
+    _br.sub_opcode = SUB_OPCODE_BR
+    _br.opcode = OPCODE_BRANCH
+    return _br.all
+
+
 def i_jumpr(offset, threshold, condition):
     offset = get_rel(offset)
     threshold = get_imm(threshold)
     condition = get_cond(condition)
     if condition == 'lt':
-        cmp_op = B_CMP_L
+        cmp_op = BRCOND_LT
     elif condition == 'ge':
-        cmp_op = B_CMP_GE
+        cmp_op = BRCOND_GE
+    elif condition == 'le':
+        threshold += 1
+        cmp_op = BRCOND_LT
+    elif condition == 'gt':
+        threshold += 1
+        cmp_op = BRCOND_GE
+    elif condition == 'eq':
+        # jump of next JUMPR
+        walk = _jump_relr(threshold + 1, BRCOND_GE, 2)
+        # jump to target
+        offset -= 1  # adjust for the additional JUMPR instruction
+        result = _jump_relr(threshold, BRCOND_GE, offset)
+        return (walk, result)
     else:
         raise ValueError("invalid comparison condition")
-    _b.imm = threshold
-    _b.cmp = cmp_op
-    _b.offset = abs(offset)
-    _b.sign = 0 if offset >= 0 else 1
-    _b.sub_opcode = SUB_OPCODE_B
-    _b.opcode = OPCODE_BRANCH
-    return _b.all
+    return _jump_relr(threshold, cmp_op, offset)
+
+
+def _jump_rels(threshold, cond, offset):
+    """
+    Equivalent of I_JUMP_RELS macro in binutils-esp32ulp
+    """
+    _bs.imm = threshold
+    _bs.cmp = cond
+    _bs.offset = abs(offset)
+    _bs.sign = 0 if offset >= 0 else 1
+    _bs.sub_opcode = SUB_OPCODE_BS
+    _bs.opcode = OPCODE_BRANCH
+    return _bs.all
 
 
 def i_jumps(offset, threshold, condition):
     offset = get_rel(offset)
     threshold = get_imm(threshold)
     condition = get_cond(condition)
     if condition == 'lt':
-        cmp_op = BC_CMP_LT
-    elif condition == 'gt':
-        cmp_op = BC_CMP_GT
-    elif condition == 'eq':
-        cmp_op = BC_CMP_EQ
+        cmp_op = BRCOND_LT
+    elif condition == 'le':
+        cmp_op = BRCOND_LE
+    elif condition == 'ge':
+        cmp_op = BRCOND_GE
+    elif condition in ('eq', 'gt'):
+        if condition == 'eq':
+            walk_cond = BRCOND_LT
+            result_cond = BRCOND_LE
+        elif condition == 'gt':
+            walk_cond = BRCOND_LE
+            result_cond = BRCOND_GE
```

Comment:
hmm, naming seems unintuitive to me.

how about using skip_cond (instead of walk_cond), because the 2nd jump instruction is skipped?

how about using jump_cond (instead of result_cond), because that is the condition in the 2nd jump instruction that eventually really jumps to the target?

fix globally...

Reply:
Very good point. I actually didn't think about this too much - I simply used the names that espressif used in binutils-esp32ulp. I do remember them not making much sense to me there, but didn't stop to think about changing them in our implementation :).

I really like your suggestions for names.